presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -273,9 +278,9 @@ public void flushCache()
     }
 
     @Override
-    public Optional<Table> getTable(String databaseName, String tableName)
+    public Optional<Table> getTable(HiveContext context, String databaseName, String tableName)
     {
-        return get(tableCache, hiveTableName(databaseName, tableName));
+        return get(tableCache, hiveTableNameWithContext(context, databaseName, tableName));
```

Comment:
This needs to be cached individually, separately for every user, when impersonation is enabled, but it should be cached once when it is not. Currently, it's always cached individually.

Reply:
That would do. Note that you then also need to change downstream interfaces (eg BridgingMetastore, etc) so that they take `Optional<>` as well. 
Since CachingM implements the same interface, it would also need to take the context as `Optional`.
This isn't going to be very readable though.

Maybe we make user in the context class optional instead of this?
We would pass "null HC object" downstream, which would then be checked on user retrieval from the context class.

Reply:
Sorry, I couldn't imagine the final structure. Is changing the ctor of CachingHiveMetastore to take ThriftHiveMetastoreConfig (impersonationEnabled) not sufficient?

Reply:
> Do you mean updating HiveMetastore interface?

I'd prefer not. I suggest relaxing our type safety a bit.

- instead of using `Optional<HiveIdentity>` in HiveMetastore interface ...
- ... use (as it is now) `HiveIdentity`, but make the `HiveIdentity#user`  `Optional`

CachingHM will need to be aware whether delegate uses impersonation. (unfortunately)
When impersonation is enabled -- everything as you impl'd.
When it's not -- instead of constructing a new `HiveIdentity` from session, we will just `new HiveIdentity(Optional.empty())` (singleton).

@ebyhr  @electrum  WDYT?

Reply:
Got it, thanks for the detailed explanation. I think there's an option to replace `identity` object in CachingHM instead of changing all `new HiveIdentity(*)`.  Which is better? I don't hesitate to replace all if the area is limited, but unfortunately even util class (ThriftMetastoreUtil) depends on it.

Reply:
We could do a variant of what I suggested in `CachingHiveMetastore`. Each method would start like this:
```java
public Optional<Table> getTable(HiveIdentity identity, String databaseName, String tableName)
{
    identity = updateIdentity(identity);
    return get(tableCache, new WithIdentity<>(identity, hiveTableName(databaseName, tableName)));
}
```

And the method would look like
```java
private HiveIdentity(HiveIdentity identity)
{
    // remove identity if not doing impersonation
    return impersonationEnabled ? identity : new HiveIdentity();
}